app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt

```diff
@@ -26,11 +37,36 @@ class TreeTrackerActivity : ComponentActivity() {
         } else {
             languageSwitcher.applyCurrentLanguage(this)
         }
-
-        setContent {
-            CustomTheme {
-                Root(viewModelFactory)
+        if (hasGPSDevice(this)) {
+            setContent {
+                CustomTheme {
+                    Root(viewModelFactory)
+                }
+            }
+        } else {
+            setContent {
+                Surface(color = AppColors.Gray) {
+                    CustomTheme {
+                        NoGPSDeviceDialog(onPositiveClick = { finishAndRemoveTask() })
+                    }
+                }
             }
         }
     }
+
+    fun hasGPSDevice(context: Context): Boolean {
+        val mgr = context.getSystemService(LOCATION_SERVICE) as LocationManager
+        val providers = mgr.allProviders
+        return providers.contains(LocationManager.GPS_PROVIDER)
+    }
```

Comment:
Let's try to keep the activity as clean as possible. It's easy for activities to turn into super large classes as time goes on. We can move this logic into one of the Location classes we have.